docs/proposals/2170-kubeflow-training-v2/README.md

```diff
@@ -0,0 +1,1431 @@
+
+```golang
+type PodSpecOverride struct {
+    // Name of the training replica in the training runtime template to override
+    TargetReplicatedJobs []string `json:"targetReplicatedJobs"`
+
+    // Override parameters for Containers.
+    Containers []Container `json:"container,omitempty"`
+
+    // Override parameters for InitContainers.
+    InitContainer []Container `json:"initContainer,omitempty"`
+
+    // Override parameters for volumes.
+    Volumes []corev1.Volume `json:"volume,omitempty"`
+
+    // Custom Service Account
+    ServiceAccountName string `json:"serviceAccountName,omitempty"`
+}
+
+// Override for each container.
+// Parameters from Trainer, DatasetConfig, and ModelConfig will take precedence.
+type Container struct {
+
+    // Name for the container.
+    Name string `json:"name"`
+
+    // Command for the container.
+    Command []string `json:"command,omitempty" protobuf:"bytes,3,rep,name=command"`
+
+    // Args for the container.
+    Args []string `json:"args,omitempty"`
+
+    // Env for the container.
+    Env []corev1.EnvVar `json:"env,omitempty"`
+
+    // Env for the container.
+    EnvFrom []corev1.EnvFromSource `json:"envFrom,omitempty"`
+
+    // Override parameters for volume mounts.
+    VolumeMounts []VolumeMount `json:"volumeMounts,omitempty"`
+}
+```
+
+#### Example of TrainJob with Overrides
+
+This example shows how to override user-identity for sidecar container and add volume to the
+trainer container.
+
+```yaml
+apiVersion: kubeflow.org/v2alpha1
+kind: TrainJob
+metadata:
+  name: pytorch-distributed
+  namespace: tenant-alpha
+spec:
+  trainingRuntimeRef:
+    name: pytorch-distributed-gpu
+  trainer:
+    image: docker.io/custom-training
+  podSpecOverrides:
+    - targetReplicatedJobs:
+        - initializer
+          node
+      containers:
+        - name: user-identity
+          value: 123
+        - name: trainer
+          volumeMounts:
+            - name: user-123-volume
+              mountPath: /workspace
+      volumes:
+        - name: user-123-volume
+          persistentVolumeClaim:
+            claimName: user-123-volume
+```
+
+## The Training Runtime API
+
+The `TrainingRuntime` is the pre-created configurations of model training on the cluster,
+representing as blueprints. For example, Elastic PyTorch training, MPI DeepSpeed configuration,
+BERT LLM Fine-Tuning.
+
+These blueprints can be deployed within the Training Operator control plane and stored in a Kubeflow
+public repository that users can apply to their clusters.
+
+Platform or ML engineers can tweak existing blueprints, based on their requirements. For example,
+using custom configurations.
+
+The Kubeflow Training Operator can maintain more Training Runtimes when the community is ready to
+support them. For example, runtimes for [Jax](https://jax.readthedocs.io/en/latest/index.html) or
+[MLX](https://ml-explore.github.io/mlx/build/html/index.html). Initially, we will support: PyTorch,
```

Comment:
The non goal states that we will not support Tensorflow, XGBoost and PaddlePaddle.

Reply:
I can add it to the non-goals, but after initial implementation with PyTorch we are planning to add those runtimes.
WDYT @kubeflow/wg-training-leads ?

Reply:
### Non-Goals

- Support MPI V1 implementation.
- Distributed training for TensorFlow, XGboost, JAX, and PaddlePaddle will be added after initial
  implementation.
- Migrate Kubeflow V1 controller to use `JobSet`.

I just mean you mention it as a nongoal already.

Reply:
Oh, yeah. Let me just add the message that these runtimes will be added after initial implementation.